sketch/Explorer_ESP_WiFi-localization_v1.2.ino

```diff
@@ -0,0 +1,162 @@
+#include <TheThingsNetwork.h>
+
+#define HEX_CHAR_TO_NIBBLE(c) ((c >= 'a') ? (c - 'a' + 0x0A) : (c - '0'))
+#define HEX_PAIR_TO_BYTE(h, l) ((HEX_CHAR_TO_NIBBLE(h) << 4) + HEX_CHAR_TO_NIBBLE(l))
+
+// Parameters for the esp stream/buffer
+#define SF_ESP8266_DEFAULT_WAIT 5000
+#define SF_ESP8266_BUFFER_SIZE 128
+#define ESP_SERIAL_SPEED 9600
+#define ACCESS_POINTS 3
+#define SF_WIFI_BSSID_SIZE 6
+
+// Set your AppEUI and AppKey
+const char *appEui = "0000000000000000";
+const char *appKey = "00000000000000000000000000000000";
+
+// Replace REPLACE_ME with TTN_FP_EU868 or TTN_FP_US915
+#define freqPlan REPLACE_ME
+
+// Define the used serials
+#define debugSerial SerialUSB
+#define loraSerial Serial2
+#define espSerial Serial
+
+// Prototyping the functions
+char *readLine(uint32_t waitTime = SF_ESP8266_DEFAULT_WAIT);
+bool waitForOK(uint32_t waitTime = SF_ESP8266_DEFAULT_WAIT);
+void sendBssid();
+
+// Declaring variables
+uint8_t payloadLength = ACCESS_POINTS * SF_WIFI_BSSID_SIZE;
+char buffer[SF_ESP8266_BUFFER_SIZE];
+byte aps[ACCESS_POINTS][SF_WIFI_BSSID_SIZE];
+
+// TTN constructor
+TheThingsNetwork ttn(loraSerial, debugSerial, freqPlan);
+
+void setup()
+{
+  // Wait a maximum of 10s for Serial Monitor
+  while (!debugSerial && millis() < 10000)
+    ;
+
+  loraSerial.begin(57600);
+  debugSerial.begin(ESP_SERIAL_SPEED);
+  espSerial.begin(ESP_SERIAL_SPEED);
+
+  debugSerial.println("-- STATUS");
+  ttn.showStatus();
+
+  debugSerial.println("-- JOIN");
+  ttn.join(appEui, appKey);
+}
+
+void loop()
+{
+  uint8_t bssidNumber = 0;
+  uint32_t timeout;
+  // command to esp, set station mode
+  espSerial.println(F("AT+CWMODE=1"));
+  bool x = waitForOK();
+
+  if (!x)
+  {
+    debugSerial.println(F("Set mode failed"));
+    goto exit;
+  }
+
+  // Command to esp to get accesspoint information
+  espSerial.println(F("AT+CWLAP"));
+
+  char *line;
+  timeout = millis() + 3000;
+
+  while (millis() < timeout && bssidNumber < ACCESS_POINTS)
+  {
+    line = readLine();
+    debugSerial.println(line);
+    if (strncmp("OK", line, 2) == 0)
+    {
+      debugSerial.println(F("end of access points"));
+      break;
+    }
+    if (strncmp("+CWLAP", line, 2) != 0)
+    {
+      delay(1);
+      debugSerial.println(F("No Bssid"));
```

Comment:
Does this happen?

Reply:
What happens in the cases i tested it is as follows: Between the echo of `+CWLAP` and the return of the first bssid, there is a empty line. It reacts to this empty line. Which is not really useful. But i think it is useful when there are no wifi signals nearby.

Reply:
So it prints this line always, because there's always an empty line?

Turn this then into a message that the user understands, e.g. `No access points found`, when there are indeed no access points found – not always

Reply:
yeah everytime it finds BSSID's it gives this 
![image](https://user-images.githubusercontent.com/24267554/33615349-a027ec6e-d9e2-11e7-8c7b-cf2c56327157.png)
The first time it gives a good description
but the second time it's not quite correct. i'll fix this

Reply:
fixed it